Add PR number to version string

rotorman opened the pull request:
WiP - if it works, I will find out after the checks complete...
Thus please do not merge yet...

pfeerick commented:
Oh, very nice... from the build log that looks like attempt 2 did it ;) Then should be able to do something like v2.8.0-PR1887 or something instead of merge :)

btw, for the tags, am I right in thinking that for 2.8.0, since the tag for say RC1 will be `v2.8.0-rc1`, it is going to say `v2.8.0-v2.8.0-rc1`? Maybe it replace the version tag when it's a tagged version?

![image](https://user-images.githubusercontent.com/5500713/164970069-c4db4c26-aeef-4e1a-89c0-0e72f10b9acc.png)

Also, I just realised another side effect of this... altering the suffix also propagates through to `semver` flag in `radio.yml`... so do we actually want to alter the version number, or add another tag (like in `stamp.h.in`) for the suffix, and use that when present in the UI? i.e. VERSION_SUFFIX would get used when present for PRs, branches, self-builds, VERSION_TAG would get used when a tagged build is detected... something like that.

rotorman commented:
> Oh, very nice... from the build log that looks like attempt 2 did it ;) Then should be able to do something like v2.8.0-PR1887 or something instead of merge :)

Unfortunately it does not work (yet) - still only `-merge` suffix:

![grafik](https://user-images.githubusercontent.com/21011587/164973375-bc632238-d5b9-4582-80e0-e24d8f64ceac.png)

> btw, for the tags, am I right in thinking that for 2.8.0, since the tag for say RC1 will be `v2.8.0-rc1`, it is going to say `v2.8.0-v2.8.0-rc1`? Maybe it replace the version tag when it's a tagged version?

> Also, I just realised another side effect of this... altering the suffix also propagates through to `semver` flag in `radio.yml`... so do we actually want to alter the version number, or add another tag (ie. in `stamp.h.in`) for the suffix, and use that when present in the UI? i.e. VERSION_SUFFIX would get used when present for PRs, branches, self-builds, VERSION_TAG would get used when a tagged build is detected... something like that.

Indeed, I will look into this, as this is a good idea to split the suffix into own variable.

rotorman commented:
I hope to have addressed all issues.

This is the result, when flashed on physical TX16S mkII with the GitHub build from this PR:

![grafik](https://user-images.githubusercontent.com/21011587/164984641-6593c184-d223-4c3e-8ec9-f3ecb918ae0f.png)

The same sourcetree, built locally and flashed on TX16S mkII:

![grafik](https://user-images.githubusercontent.com/21011587/164984664-69af89a0-be86-41a4-a946-b1596f377f7e.png)

Bootloader flashed with the binary built by this PR in GitHub:

![grafik](https://user-images.githubusercontent.com/21011587/164984718-eb1efe63-818c-4418-a04c-f1b06949778e.png)

Bootloader display, when selecting the firmware binary from this PR:

![grafik](https://user-images.githubusercontent.com/21011587/164984756-4e6c1ca8-dfdb-48bd-b553-edaab193e9e5.png)

Bootloader display, when selecting the self-built binary:

![grafik](https://user-images.githubusercontent.com/21011587/164984796-e90aac98-44b9-48a7-bc0e-6022c4ac5ae5.png)

I have not yet tested nightly (I guess possible only after this PR is merged) or the tagged builds (could do a test tag and then delete it later).

rotorman commented:
The output on RadioMaster Zorro with this PR build:

![grafik](https://user-images.githubusercontent.com/21011587/164987469-5f11f031-0673-4b0a-af3b-6b3b05f6c55e.png)

the same codebase, but locally built:

![grafik](https://user-images.githubusercontent.com/21011587/164987475-a0af2251-8977-4aff-b01c-635feb4928a4.png)

Bootloader flashed from this PR:

![grafik](https://user-images.githubusercontent.com/21011587/164987517-3d34d94b-ad27-4c1b-9f40-a18f3e880281.png)

Selecting firmware limited to version string only due to small screen size:

![grafik](https://user-images.githubusercontent.com/21011587/164987536-e66b218e-d8d6-42d6-8fda-ed7bb3b26b05.png)

Eldenroot commented:
the whole text should be visible - maybe add a posibility to select it and click on it to see all the info?

rotorman commented:
Ideally, could be scrolling text, if the text to be displayed is longer than the display width, but this is an idea for a further PR.

Eldenroot commented:
Scrolling would be nice... and I think it should be bundled into one PR, why to create another one? Now it is only a partially done.

pfeerick commented:
I don't know if any changes to the bootloader are warranted... long lines have always been a problem there and enough info is conveyed for it without increasing the complexity of it's code.

For the long version string (which mainly only affects 128x64 B&W screens since they don't have enough width), I've moved the commit SHA to it's own line and it's only shown if it is a non-tagged version (since if it's tagged, it's commit hash is tied to its tag).

pfeerick reviewed (approved):
Fantastic work Risto :) We can certainly push a tag to test it - although I technically the nightly build *should* trigger that is it is supposed to be a moving tag if I remember the github build process properly. Which now that of I think of *that*, it makes me think I need to re-add the git hash for tagged B&W builds... I guess we'll find out soon! :facepalm:
